src/r1cs/sparse.rs

```diff
@@ -118,6 +120,43 @@ impl<F: PrimeField> SparseMatrix<F> {
       .collect()
   }
 
+  /// Multiply by a witness representing a dense vector; uses rayon/gpu.
+  pub fn multiply_witness(&self, W: &[F], u: &F, X: &[F]) -> Vec<F> {
+    assert_eq!(self.cols, W.len() + X.len() + 1, "invalid shape");
+
+    self.multiply_witness_unchecked(W, u, X)
+  }
+
+  /// Multiply by a witness representing a dense vector; uses rayon/gpu.
+  /// This does not check that the shape of the matrix/vector are compatible.
+  #[tracing::instrument(
+    skip_all,
+    level = "trace",
+    name = "SparseMatrix::multiply_vec_unchecked"
+  )]
+  pub fn multiply_witness_unchecked(&self, W: &[F], u: &F, X: &[F]) -> Vec<F> {
+    let num_vars = W.len();
+    // preallocate the result vector
+    let mut result = Vec::with_capacity(self.indptr.len() - 1);
```

Comment:
This isn't allocation-free? I think the goal here is to be explicit ( in the PR summary, and ideally in a comment here) on how this performs less allocations by avoiding the explicit concatenation of the witness vector's terms.